docs: add info about tailcalls in bpf subprograms (#13888)
cilium-team/cilium: Documentation/bpf.rst

A caller can pass pointers (e.g. to the caller's stack frame) down to the
callee, but never vice versa.

BPF JIT compilers emit separate images for each function body and later fix
up the function call addresses in the image in a final JIT pass. This has
proven to require minimal changes to the JITs in that they can treat BPF to
BPF calls as conventional BPF helper calls.

Up to kernel 5.9, BPF tail calls and BPF subprograms excluded each other. BPF

> programs that utilized tail calls couldn't take the benefit of shrinked program

Reviewer: nit: `benefit of shrinked program` -> `benefit of reducing program image size`

Reviewer: Please also run `./Documentation/update-spelling_wordlist.sh func unwinded` so the latter two get added to the word list and don't cause doc CI failure. Thanks for extending the doc!

images and faster load times. Linux kernel 5.10 finally allows user to take the
best of two worlds and adds the ability to combine the BPF subprograms with tail

> calls.

Reviewer 2: ```suggestion
images and faster load times. Linux kernel 5.10 finally allows users to bring the best of two worlds and adds the ability to combine the BPF subprograms with tail calls.
```

This improvement comes with some restrictions, though. Mixing these two features

> could cause kernel stack overflow. To get an idea of what might happen, see the

Reviewer 2: ```suggestion
This improvement comes with some restrictions, though. Mixing these two features
can cause a kernel stack overflow. To get an idea of what might happen, see the
```

picture below that illustrates the mix of bpf2bpf calls and tail calls:

.. image:: images/bpf_tailcall_subprograms.png
    :align: center

Tail call, before the actual jump to target program, will unwind only its
current stack frame. As we can see on the example above, if a tail call takes
place from within the sub function, the functions's (func1) stack frame will be
present on stack when a program execution is at func2. Once the final (func3)
function is terminated, all the previous stack frames will be unwinded and

> control will get back to the caller of BPF program.

Reviewer 2: ```suggestion
Tail calls, before the actual jump to the target program, will unwind only its
current stack frame. As we can see in the example above, if a tail call occurs
from within the sub-function, the functions' (func1) stack frame will be
present on the stack when program execution is at func2. Once the final function(func3)
terminates, all the previous stack frames will be unwinded, and
control will get back to the BPF program caller.
```

Additional logic was introduced for detecting this feature combination and in
that case it limits the stack size throughout the whole call chain down to
256 bytes per each subprogram (note that if verifier detects the bpf2bpf call
then main function is treated as a sub function as well). In total, with that
restriction, BPF program's call chain can consume at most 8k of stack space.
This limit comes from the 256 bytes per stack frame multiplied by 32 which is
the tail call count limit. Without that, normally BPF programs can operate on
512 byte stack size which would yield the 16k size in total for maximum count of
tail calls and that solely would overflow the stack on some architectures.

One more thing to mention is that this feature combination is currently

> supported only on x86-64 architecture.

Reviewer 2: How about rewording this to something like?
```
The kernel introduced additional logic for detecting this feature combination.
There is a limit on the stack size throughout the whole call chain down to 256 bytes
per subprogram (note that if the verifier detects the bpf2bpf call, then the main
function is treated as a sub-function as well). In total, with this restriction,
the BPF program's call chain can consume at most 8KB of stack space.
This limit comes from the 256 bytes per stack frame multiplied by the tail call count limit(32).
Without this, the BPF programs will operate on 512-byte stack size, yielding the
16KB size in total for the maximum count of tail calls that would overflow the
stack on some architectures.
One more thing to mention is that this feature combination is currently supported only on the x86-64 architecture.
```
